### PR TITLE
Fancy lighting

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request
Third times the charm?
## Why It's Good For The Game
Looks cool
Testing with players was apparently fine but if it breaks again im going to try mipmapping instead

## Changelog
:cl:
refactor: Lighting rework
/:cl:

### CODE REVIEW

File: code/__DEFINES/colors.dm [hunk 1]
```diff
@@ -112,8 +112,12 @@ Some defines to generalise colours used in lighting.
 Important note: colors can end up significantly different from the basic html picture, especially when saturated
 */
 
+/// Full white. rgb(255, 255, 255)
+#define LIGHT_COLOR_WHITE		"#FFFFFF"
```
Comment (BraveMole) [inline code/__DEFINES/colors.dm:116]:
```suggestion
#define LIGHT_COLOR_WHITE "#FFFFFF"
```

File: code/__DEFINES/colors.dm [hunk 2]
```diff
@@ -112,8 +112,12 @@ Some defines to generalise colours used in lighting.
...
 /// Bright but quickly dissipating neon green. rgb(100, 200, 100)
 #define LIGHT_COLOR_GREEN "#64C864"
+/// Warm red color rgb(250, 66, 66)
+#define LIGHT_COLOR_RED        "#ff3b3b"
```
Comment (BraveMole) [inline code/__DEFINES/colors.dm:120]:
```suggestion
#define LIGHT_COLOR_RED "#ff3b3b"
```

File: code/__DEFINES/lighting.dm [hunk 3]
```diff
@@ -4,34 +4,16 @@
...
-#define LIGHTING_HEIGHT 1 // height off the ground of light sources on the pseudo-z-axis, you should probably leave this alone
-#define LIGHTING_ROUND_VALUE (1 / 64) //Value used to round lumcounts, values smaller than 1/129 don't matter (if they do, thanks sinking points), greater values will make lighting less precise, but in turn increase performance, VERY SLIGHTLY.
-
 #define LIGHTING_ICON 'icons/effects/lighting_object.dmi' // icon used for lighting shading effects
+#define LIGHTING_ICON_BIG 'icons/effects/lighting_object_big.dmi' //! icon used for lighting shading effects
 
-// If the max of the lighting lumcounts of each spectrum drops below this, disable luminosity on the lighting objects.
-// Set to zero to disable soft lighting. Luminosity changes then work if it's lit at all.
-#define LIGHTING_SOFT_THRESHOLD 0
+#define ALPHA_TO_INTENSITY(alpha) (-(((clamp(alpha, 0, 22) - 22) / 6) ** 4) + 255)
```
Comment (BraveMole) [inline code/__DEFINES/lighting.dm:15]: ew

File: code/__DEFINES/lighting.dm [hunk 4]
```diff
@@ -100,3 +70,33 @@ do { \
...
+
+//Bay lighting engine shit, not in /code/modules/lighting because BYOND is being shit about it	//thats how defines work, hello?
+#define LIGHTING_INTERVAL       5 // frequency, in 1/10ths of a second, of the lighting process
+
+#define MOVABLE_MAX_RANGE 7
+
+#define LIGHTING_FALLOFF        1 // type of falloff to use for lighting; 1 for circular, 2 for square
+#define LIGHTING_LAMBERTIAN     0 // use lambertian shading for light sources
+#define LIGHTING_HEIGHT         1 // height off the ground of light sources on the pseudo-z-axis, you should probably leave this alone
+#define LIGHTING_ROUND_VALUE    (1 / 64) //Value used to round lumcounts, values smaller than 1/129 don't matter (if they do, thanks sinking points), greater values will make lighting less precise, but in turn increase performance, VERY SLIGHTLY.
```
Comment (BraveMole) [inline code/__DEFINES/lighting.dm:83]: lot of tabs

File: code/__HELPERS/unsorted.dm [hunk 5]
```diff
@@ -970,6 +970,54 @@ GLOBAL_LIST_INIT(wallitems, typecacheof(list(
...
+/proc/dview(range = world.view, center, invis_flags = 0)
+	if(!center)
+		return
+
+	GLOB.dview_mob.loc = center
+
+	GLOB.dview_mob.see_invisible = invis_flags
+
+	. = view(range, GLOB.dview_mob)
+	GLOB.dview_mob.loc = null
```
Comment (BraveMole) [inline code/__HELPERS/unsorted.dm:985]: lel this proc is hilarious, thanks byond

File: code/__HELPERS/unsorted.dm [hunk 6]
```diff
@@ -970,6 +970,54 @@ GLOBAL_LIST_INIT(wallitems, typecacheof(list(
...
+
+/mob/dview
+	name = "INTERNAL DVIEW MOB"
+	invisibility = 101
+	density = FALSE
+	see_in_dark = 1e6
+	move_resist = INFINITY
+	var/ready_to_die = FALSE
```
Comment (BraveMole) [inline code/__HELPERS/unsorted.dm:993]: autodoc?

File: code/__HELPERS/unsorted.dm [hunk 7]
```diff
@@ -970,6 +970,54 @@ GLOBAL_LIST_INIT(wallitems, typecacheof(list(
...
+	GLOB.dview_mob.loc = center
+
+	GLOB.dview_mob.see_invisible = invis_flags
+
+	. = view(range, GLOB.dview_mob)
+	GLOB.dview_mob.loc = null
+
+/mob/dview
+	name = "INTERNAL DVIEW MOB"
+	invisibility = 101
```
Comment (BraveMole) [inline code/__HELPERS/unsorted.dm:989]: 101?

File: code/_compile_options.dm [hunk 8]
```diff
@@ -71,3 +71,5 @@
 #ifdef CITESTING
 #define TESTING
 #endif
+
+//#define SHADOW_DEBUG
```
Comment (BraveMole) [inline code/_compile_options.dm:75]: ?

File: code/controllers/subsystem/evacuation.dm [hunk 9]
```diff
@@ -6,9 +6,10 @@ SUBSYSTEM_DEF(evacuation)
 	var/pod_cooldown
 	var/evac_time
 	var/evac_status = EVACUATION_STATUS_STANDING_BY
+	var/list/alarm_lights = list()
```
Comment (BraveMole) [inline code/controllers/subsystem/evacuation.dm:9]: autodoc

File: code/controllers/subsystem/lighting.dm [hunk 10]
```diff
@@ -2,86 +2,93 @@ SUBSYSTEM_DEF(lighting)
...
 	wait = 2
 	init_order = INIT_ORDER_LIGHTING
-	flags = SS_TICKER
-	runlevels = RUNLEVEL_LOBBY|RUNLEVEL_SETUP|RUNLEVEL_GAME|RUNLEVEL_POSTGAME
-	var/static/list/sources_queue = list() // List of lighting sources queued for update.
-	var/static/list/corners_queue = list() // List of lighting corners queued for update.
-	var/static/list/objects_queue = list() // List of lighting objects queued for update.
 
+	var/duplicate_shadow_updates_in_init = 0
+	var/total_shadow_calculations = 0
```
Comment (BraveMole) [inline code/controllers/subsystem/lighting.dm:7]: autodocs?

File: code/controllers/subsystem/lighting.dm [hunk 11]
```diff
@@ -2,86 +2,93 @@ SUBSYSTEM_DEF(lighting)
...
 /datum/controller/subsystem/lighting/fire(resumed, init_tick_checks)
 	MC_SPLIT_TICK_INIT(3)
 	if(!init_tick_checks)
 		MC_SPLIT_TICK
-	var/list/queue = sources_queue
-	var/i = 0
-	for(i in 1 to length(queue))
-		var/datum/light_source/L = queue[i]
+	var/updators_num = 0
+	for(updators_num in 1 to static_sources_queue.len)
```
Comment (BraveMole) [inline code/controllers/subsystem/lighting.dm:34]:
```suggestion
	for(updators_num in 1 to length(static_sources_queue))
```

even though static_source being null is probably impossible

File: code/controllers/subsystem/lighting.dm [hunk 12]
```diff
@@ -2,86 +2,93 @@ SUBSYSTEM_DEF(lighting)
...
+	if(updators_num)
+		static_sources_queue.Cut(1, ++updators_num)
+		updators_num = 0
 	if(!init_tick_checks)
 		MC_SPLIT_TICK
 
-	queue = corners_queue
-	for(i in 1 to length(queue))
-		var/datum/lighting_corner/C = queue[i]
+	for(updators_num in 1 to corners_queue.len)
```
Comment (BraveMole) [inline code/controllers/subsystem/lighting.dm:50]:
```suggestion
	for(updators_num in 1 to length(corners_queue))
```

File: code/controllers/subsystem/lighting.dm [hunk 13]
```diff
@@ -2,86 +2,93 @@ SUBSYSTEM_DEF(lighting)
...
+	if(updators_num)
+		corners_queue.Cut(1, ++updators_num)
+		updators_num = 0
 	if(!init_tick_checks)
 		MC_SPLIT_TICK
 
-	queue = objects_queue
-	for(i in 1 to length(queue))
-		var/atom/movable/lighting_object/O = queue[i]
+	for(updators_num in 1 to objects_queue.len)
```
Comment (BraveMole) [inline code/controllers/subsystem/lighting.dm:65]:
```suggestion
	for(updators_num in 1 to length(objects_queue))
```

File: code/controllers/subsystem/lighting.dm [hunk 14]
```diff
@@ -2,86 +2,93 @@ SUBSYSTEM_DEF(lighting)
...
+		else if (MC_TICK_CHECK)
 			break
+	if(updators_num)
+		objects_queue.Cut(1, ++updators_num)
+	if(!init_tick_checks)
+		MC_SPLIT_TICK
 
-	if(i)
-		queue.Cut(1, i + 1)
+	for(updators_num in 1 to mask_queue.len)
```
Comment (BraveMole) [inline code/controllers/subsystem/lighting.dm:81]:
```suggestion
	for(updators_num in 1 to length(mask_queue))
```

File: code/modules/lighting/lighting_area.dm [hunk 15]
```diff
@@ -1,32 +1,55 @@
...
+/area/proc/update_base_lighting()
+	if(!area_has_base_lighting && (!base_lighting_alpha || !base_lighting_color))
+		return
+
+	if(area_has_base_lighting)
+		remove_base_lighting()
+		if(base_lighting_alpha && base_lighting_color)
+			add_base_lighting()
+	else
+		add_base_lighting()
```
Comment (BraveMole) [inline code/modules/lighting/lighting_area.dm:37]: early return

File: code/modules/lighting/lighting_atom.dm [hunk 16]
```diff
@@ -1,48 +1,62 @@
+
 // The proc you should always use to set the light of this atom.
 // Nonesensical value for l_color default, so we can detect if it gets set to null.
 #define NONSENSICAL_VALUE -99999
-/atom/proc/set_light(l_range, l_power, l_color = NONSENSICAL_VALUE)
+/atom/proc/set_light(l_range, l_power, l_color = NONSENSICAL_VALUE, mask_type = null)
```
Comment (BraveMole) [inline code/modules/lighting/lighting_atom.dm:5]: autodoc while you are here

File: code/modules/lighting/lighting_atom.dm [hunk 17]
```diff
@@ -1,48 +1,62 @@
...
 	set waitfor = FALSE
-	if(QDELETED(src))
+	if(light_system == STATIC_LIGHT)
+		static_update_light()
 		return
 
-	if(light_system != STATIC_LIGHT)
-		CRASH("update_light() for [src] with following light_system value: [light_system]")
+	if (QDELETED(src))
+		return
```
Comment (BraveMole) [inline code/modules/lighting/lighting_atom.dm:40]: i don't get why the qdeleted check is not first

File: code/modules/lighting/lighting_atom.dm [hunk 18]
```diff
@@ -1,48 +1,62 @@
...
 
-	if(!light_power || !light_range) // We won't emit light anyways, destroy the light source.
-		QDEL_NULL(light)
+	if (!light_power || !light_range) // We won't emit light anyways, destroy the light source.
+		if(light)
+			QDEL_NULL(light)
```
Comment (BraveMole) [inline code/modules/lighting/lighting_atom.dm:44]: if ((!light_power || !light_range) && light)
   QDEL_NULL(light)
   return

File: code/modules/lighting/lighting_atom.dm [hunk 19]
```diff
@@ -1,48 +1,62 @@
...
 	else
-		if(!ismovableatom(loc)) // We choose what atom should be the top atom of the light here.
-			. = src
-		else
-			. = loc
-
-		if(light) // Update the light or create it if it does not exist.
-			light.update(.)
+		if(light && light_mask_type && (light_mask_type != light.mask_type))
+			QDEL_NULL(light)
```
Comment (BraveMole) [inline code/modules/lighting/lighting_atom.dm:47]: you could even put this check with the other one

File: code/modules/lighting/lighting_atom.dm [hunk 20]
```diff
@@ -1,48 +1,62 @@
...
+		if (!light) // Update the light or create it if it does not exist.
+			light = new /datum/dynamic_light_source(src, light_mask_type)
```
Comment (BraveMole) [inline code/modules/lighting/lighting_atom.dm:49]: early return

File: code/modules/lighting/lighting_atom.dm [hunk 21]
```diff
@@ -75,26 +89,36 @@
 		return
 	recalculate_directional_opacity()
 
+
+
+// If we have opacity, make sure to tell (potentially) affected light sources.
+/atom/movable/Destroy()
+	. = ..()
+	QDEL_NULL(light)
+	QDEL_NULL(static_light)
```
Comment (BraveMole) [inline code/modules/lighting/lighting_atom.dm:98]: return ..()

No?

File: code/modules/lighting/lighting_mask/lighting_mask.dm [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+	overlay_merger.appearance = dud.appearance
+	overlays += overlay_merger
+	dud.overlays.Cut()
+
+	//Now we are facing this direction
+	current_angle = angle
+
+///Setter proc for colors
+/atom/movable/lighting_mask/proc/set_color(colour = "#ffffff")
+	color = colour
```
Comment (BraveMole) [inline code/modules/lighting/lighting_mask/lighting_mask.dm:112]: why does this exist?

File: code/modules/lighting/lighting_mask/lighting_mask.dm [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+	if(intensity >= 0)
+		alpha = ALPHA_TO_INTENSITY(intensity)
+		blend_mode = BLEND_ADD
+	else
+		alpha = ALPHA_TO_INTENSITY(-intensity)
+		blend_mode = BLEND_SUBTRACT
+
+///The holder atom turned, spins the mask if it's needed
+/atom/movable/lighting_mask/proc/rotate_mask_on_holder_turn(new_direction)
+	rotate(dir2angle(new_direction) - 180)
```
Comment (BraveMole) [inline code/modules/lighting/lighting_mask/lighting_mask.dm:125]: should be a macro

File: code/modules/lighting/lighting_mask/shadow_calculator.dm [hunk 24]
```diff
@@ -0,0 +1,685 @@
...
+	DO_SOMETHING_IF_DEBUGGING_SHADOWS(log_game("MA_new_time: [MA_new_time]"))
+	DO_SOMETHING_IF_DEBUGGING_SHADOWS(log_game("MA_vars_time: [MA_vars_time]"))
+	DO_SOMETHING_IF_DEBUGGING_SHADOWS(log_game("overlays_add_time: [overlays_add_time]"))
+	DO_SOMETHING_IF_DEBUGGING_SHADOWS(log_game("merge_overlays_time: [mergetimereal]ms"))
+	DO_SOMETHING_IF_DEBUGGING_SHADOWS(log_game("[TICK_USAGE_TO_MS(timer)]ms to process total."))
+
+///Converts a triangle into a matrix that can be applied to a standardized triangle
+///to make it represent the points.
+///Note: Ignores translation because
+/atom/movable/lighting_mask/proc/triangle_to_matrix(list/triangle)
```
Comment (BraveMole) [inline code/modules/lighting/lighting_mask/shadow_calculator.dm:251]: bad autodoc format

**Review by BraveMole — CHANGES_REQUESTED** [review]
Yeah i'll probably never try to review the rest cause these vertex manipulation are killing me

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/lighting.dm [hunk 25]
```diff
@@ -2,86 +2,93 @@ SUBSYSTEM_DEF(lighting)
...
+		var/atom/movable/lighting_mask/mask_to_update = mask_queue[updators_num]
 
+		mask_to_update.calculate_lighting_shadows()
+		if(init_tick_checks)
+			CHECK_TICK
+		else if (MC_TICK_CHECK)
+			break
+	if(updators_num)
+		mask_queue.Cut(1, ++updators_num)
```
Comment (SplinterGP) [inline code/controllers/subsystem/lighting.dm:90]: list index out of bounds in here

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/__HELPERS/unsorted.dm [hunk 26]
```diff
@@ -970,6 +970,54 @@ GLOBAL_LIST_INIT(wallitems, typecacheof(list(
...
+		GLOB.dview_mob = new
+	return ..()
+
+
+#define FOR_DVIEW(type, range, center, invis_flags) \
+	GLOB.dview_mob.loc = center;           \
+	GLOB.dview_mob.see_invisible = invis_flags; \
+	for(type in view(range, GLOB.dview_mob))
+
+#define FOR_DVIEW_END GLOB.dview_mob.loc = null
```
[the remark this answers is not in the document]
Comment (TiviPlus) [reply code/__HELPERS/unsorted.dm:1019]: Why though this is a dview related proc

File: code/_compile_options.dm (hunk 8 above)
[the remark this answers is not in the document]
Comment (TiviPlus) [reply code/_compile_options.dm:75]: It's debug highlighting what do you want me to say it' what it says on the tin

File: code/controllers/subsystem/lighting.dm (hunk 25 above)
Comment (TiviPlus) [reply code/controllers/subsystem/lighting.dm:94]: Fixed

**Comment by Al-1ce** [comment]
no way to turn it off? fancy has always lagged me, I know a few other players too with the same issue.

**Comment by SplinterGP** [comment]
Apparently way more people than me (who has a potato PC so i am to blame) has lag too

**Comment by Al-1ce** [comment]
>it has to be seen if they werent on 514, which we need to bump for this at minimum.


this is barely english so I don't know if you were saying no one from 514 has had an issue or only people from 514 but I'm on 514 myself and still lagging

**Comment by TiviPlus** [comment]
> > it has to be seen if they werent on 514, which we need to bump for this at minimum.
> 
> this is barely english so I don't know if you were saying no one from 514 has had an issue or only people from 514 but I'm on 514 myself and still lagging.
> 
> I should also note I experienced lag on the stable version of BYOND with the fancy lighting which is what drove me to updating to 514.
> 
> 

There was a warning to update when you connect

File: code/modules/client/client_procs.dm [hunk 27]
```diff
@@ -1,7 +1,7 @@
 #define UPLOAD_LIMIT 1000000	//Restricts client uploads to the server to 1MB
 #define UPLOAD_LIMIT_ADMIN 10000000	//Restricts admin uploads to the server to 10MB
 
-#define MAX_RECOMMENDED_CLIENT 1557
+#define MAX_RECOMMENDED_CLIENT 1561
 #define MIN_RECOMMENDED_CLIENT 1526
```
Comment (BraveMole) [inline code/modules/client/client_procs.dm:5]: and this?

**Comment by Losenis** [comment]
Add a way to toggle this off, I'm getting like 5fps near fires

**Comment by Al-1ce** [comment]
> Add a way to toggle this off, I'm getting like 5fps near fires


That's not how it works. Either the entire server has it or they don't. Fancy Lighting is understandable on slower/HRP servers but not fast-paced combat focused stuff like TGMC. It slows the server down and adds nothing in terms of quality.

**Comment by Losenis** [comment]
Fires are wack
image

**Comment by TiviPlus** [comment]
Ping me in discord if you see black tiles

**Comment by Al-1ce** [comment]
unmerge my potato can't take it *warcry

**Comment by PowerfulBacon** [comment]
I worked on this for months and apart from the hard dels and runtimes causing lag in the original bee PR, I don't think byond is ready for something like this. 
It adds in way too many overlays, it can get up to hundreds for larger light sources (and I think not making the light sources tile bound causes byond to run a really laggy check to see if its in view).
The only way I can see it being done with the shadows is with changes to byond itself.
It needs a way to override the internal byond checks to see if something should be rendered, so that the lights can be rendered if their box is within view. (While typing this I was thinking, could you possible do something like with XRAY flags SEE_MOB or whatever so that you can always see lighting objects through walls, then make them tile bound?)
The other thing that would be a saviour for this lighting method is something like shader editting / a way to edit icons quickly so that instead of having a shit ton of overlays on an object, it can be a single icon where the overlay is generated in some efficient manner idk.

On bee, I've got it in an acceptable condition by adopting the TG overlay lighting approach.
Spinning lighting is still supported, direction lighting is still supported, fire flicker is still supported, however shadows are completely stripped out and the light is only rendered if you can see the source.
For the most part, it works pretty well there is a lot less lag and the effect of lights suddenly appearing isn't too noticeable unless the source is huge.
You could potentially use it for smaller lights, add TILE_BOUND to the lighting and see if that helps with the client lag, then you can have the lights update fast and have the shadows; however the major downside to that is the lights can't move smoothly with the moving object because the shadows dont have a way to animate between shadow states like in SS14.

It's not really a problem with peoples PCs not being able to handle it, its a problem with byond not being optimised / made for something working in this way.

https://github.com/BeeStation/BeeStation-Hornet/pull/5101

I experimented for so long with these and tried everything from overlays, filters, icon manipulation but everything is just simply too slow on the server, or too slow on the renderer.
The only thing I could see being quick in byond is baking every single light into an icon, however it is completely unreasonable to bake every single possible light source and shadow config since you would need millions of icons 😔

**Comment by Al-1ce** [comment]
I don't really think there is much of a point to this. It doesn't really change how many things look and just slows down the server.

**Comment by BraveMole** [comment]
> I don't really think there is much of a point to this. It doesn't really change how many things look and just slows down the server.

I don't think you have any data on how this slows or doesn't slow server? Have you really upgraded to the right version? Because you're basically the only one that still complains about performance

**Comment by TiviPlus** [comment]
> It's not really a problem with peoples PCs not being able to handle it, its a problem with byond not being optimised / made for something working in this way.

Yea agreed lummox even admits byond's rendering isnt very well done and I tried what you stated as well even going as far as to trying to prebake icons in rust code 

As for the shadows things its less the size of them and relly just the massive quantities they're present in
I have some ideas to try minimize them by more selectively rendering them using render target/source but with how large this pr is I'll try to implement that after
I'm also know that this should be more viable with 515 since lummox said that he's fast tracking some more rendering features so hopefully those are useful as well
In the meanwhile though they work fine for centerpieces

**Comment by SplinterGP** [comment]
Kaboom

**Comment by Losenis** [comment]
>Because you're basically the only one that still complains about performance

Bro incen HE straight up tanks my game to single digits FPS, just because you don't hear it doesn't mean it's not happening

**Comment by Losenis** [comment]
>Merged
Aight guess I'm fucking kicked out of anything with fire in a 2 screen radius, that's fucking cool

**Comment by Al-1ce** [comment]
Can't play here anymore with this PR.

**Comment by Al-1ce** [comment]
> > I don't really think there is much of a point to this. It doesn't really change how many things look and just slows down the server.
> 
> I don't think you have any data on how this slows or doesn't slow server? Have you really upgraded to the right version? Because you're basically the only one that still complains about performance

Do I as a regular player have multiple charts and elaborate diagrams showing in detail how this slows the server? No. Are you pretending to be daft or something? I'm going by my experience - as well as other people I've talked to in the community who have experienced the same problems in limited or hightened capacity. I can **fucking guarentee** you that I'm not the only person experiencing the issues, as Losenis above has chipped in as well. I _am_ however one of the few people in the community who is willing to talk about the issues with this PR on the github, which again, doesn't represent the entirety of the community. 

I'm not sure how you could miss this so I'm going to assume you did it intently knowing one of your buddies would merge it regardless, as with most PR's here. Majoriety of the server doesn't even know what the github is or how it works, but many people were upset in OOC when the lighting addition(s) were first added to begin with. You're blind to the problem or don't give a shit, and I'm crossed on believing the second option. Whatever. I don't play on severs with fancy lighting, and I won't be playing here anymore because of it.

**Comment by BraveMole** [comment]
You are saying it is slowing down there sever box. It is not.

You are saying a lot of people complained ooc about fancy lights, when it was first introduced. Okay, but the code was changed and the identified bug and perfs drained were fixed.

Right now, with the pr merge, I don't see anyone complaining about fancy.

You didn't answer the question, what byond version are you?

And you're saying you don't play anymore with fancy, does that mean you never tested the new version with all the bug fixes, that is currently merged?

**Comment by Losenis** [comment]
It's straight up fucking unplayable with the game tanking to like 3fps. I can't fucking play like this and this shitshow got merged because 'muh graphics' on a fucking BYOND game.
image

**Comment by february-ops** [comment]
Can corfirm also slows my game, itd beautifull and to do this in byond is amazing but this is combat server and lag and this makes a lot of lag unfortunately

**Comment by Losenis** [comment]
>Right now, with the pr merge, I don't see anyone complaining about fancy in OOC.

absence of evidence etc

image
image

**Comment by Al-1ce** [comment]
In my experience, both from the initial TM and even before this was actually merged both time dilation and overall preformance for the server seemingly tanked. Regardless of the bugs that were fixed, you can't polish a turd into a diamond. I'm not trying to be crude or demeaning but that's just how BYOND and lighting seems to work. As PowerfulBacon mentioned above, BYOND is not optimised or made for something working in this way.

Right now, with the current merge just hopping back in the lobby I saw people upset over OOC about the sudden lag. I've talked to several players IRC who haven't played in numerous months, and despite their BYOND being the most recent version things are suddenly "much more slow"

I didn't answer your question about my BYOND version because it isn't relevant to the core structural issue with this - fancy lighting in an engine which is older than the technology & code to create this lighting to begin with. BYOND _at the core_ is not designed to support advanced or fancy lighting. It just isn't.

Again, I'm not sure how hard I have to bang it in but there is a **fundamental** issue with this PR and fancy lighting as a whole.

 If you're absolutely curious, I'm running the current beta version of BYOND whatever that is. Through what I've tested, swapping between the stable and beta version has no effect, and fancy lighting was as far as I can observe, is the only cause for lag.

Yes. I've played with the "bug fixes". I did when I came back to find people upset in OOC about the PR yesterday. You can fix as many bugs as you'd like with this PR and perfect the code, but there is a _**FUNDAMENTAL**_ issue with the Fancy Lighting.

As you can see from the people above, again, I'm not "basically the only one that still complains about performance".

**Comment by Al-1ce** [comment]
In my experience, both from the initial TM and even before this was actually merged both time dilation and overall preformance for the server seemingly tanked. Regardless of the bugs that were fixed, you can't polish a turd into a diamond. I'm not trying to be crude or demeaning but that's just how BYOND and lighting seems to work. As PowerfulBacon mentioned above, BYOND is not optimised or made for something working in this way.

Right now, with the current merge just hopping back in the lobby I saw people upset over OOC about the sudden lag. I've talked to several players IRC who haven't played in numerous months, and despite their BYOND being the most recent version things are suddenly "much more slow"

I didn't answer your question about my BYOND version because it isn't relevant to the core structural issue with this - fancy lighting in an engine which is older than the technology & code to create this lighting to begin with. BYOND _at the core_ is not designed to support advanced or fancy lighting. It just isn't.

Again, I'm not sure how hard I have to bang it in but there is a **fundamental** issue with this PR and fancy lighting as a whole.

 If you're absolutely curious, I'm running the current beta version of BYOND whatever that is. Through what I've tested, swapping between the stable and beta version has no effect, and fancy lighting was as far as I can observe, is the only cause for lag.

Yes. I've played with the "bug fixes". I did when I came back to find people upset in OOC about the PR yesterday. You can fix as many bugs as you'd like with this PR and perfect the code, but there is a _**FUNDAMENTAL**_ issue with the Fancy Lighting.

As you can see from the people above, again, I'm not "basically the only one that still complains about performance".

**Comment by Al-1ce** [comment]
In my experience, both from the initial TM and even before this was actually merged both time dilation and overall preformance for the server seemingly tanked. Regardless of the bugs that were fixed, you can't polish a turd into a diamond. I'm not trying to be crude or demeaning but that's just how BYOND and lighting seems to work. As PowerfulBacon mentioned above, BYOND is not optimised or made for something working in this way.

Right now, with the current merge just hopping back in the lobby I saw people upset over OOC about the sudden lag. I've talked to several players IRC who haven't played in numerous months, and despite their BYOND being the most recent version things are suddenly "much more slow"

I didn't answer your question about my BYOND version because it isn't relevant to the core structural issue with this - fancy lighting in an engine which is older than the technology & code to create this lighting to begin with. BYOND _at the core_ is not designed to support advanced or fancy lighting. It just isn't.

Again, I'm not sure how hard I have to bang it in but there is a **fundamental** issue with this PR and fancy lighting as a whole.

 If you're absolutely curious, I'm running the current beta version of BYOND whatever that is. Through what I've tested, swapping between the stable and beta version has no effect, and fancy lighting was as far as I can observe, is the only cause for lag.

Yes. I've played with the "bug fixes". I did when I came back to find people upset in OOC about the PR yesterday. You can fix as many bugs as you'd like with this PR and perfect the code, but there is a _**FUNDAMENTAL**_ issue with the Fancy Lighting.

As you can see from the people above, again, I'm not "basically the only one that still complains about performance".

**Comment by Hekzder** [comment]
Adding on to this for more anecdotes, noticeable performance decrease since Fancy Lighting is merged that is generally linked to amount of flames on the screen. I am on the latest Byond version and generally have a pretty solid rig, the performance hits for _me_ aren't terrible or crippling in most circumstances but are noticeable and it has gotten me killed on one occasion so far. Incend OBs are serious pain though and are a massive performance hit. 

Generally speaking I just don't understand this PR at all. The added effect/up side seems incredibly minimal, the Ripley and SD spinning lights are cool but how is that (or anything else) worth the performance impact? I can only imagine what this is doing to people on low end rigs. 

> Right now, with the pr merge, I don't see anyone complaining about fancy in OOC.

Brave you're literally coping my man, don't know what to tell you. I've been seeing people complain about Fancy Lighting very frequently the past few days.

**Comment by Losenis** [comment]
Fire isn't the only issue either. For some reason I'm lagging extremely hard in this particular spot.
image

Code an "off" button so I can opt out of this mandatory 'merged while people were crashing to fire' shenanigans.

**Comment by Al-1ce** [comment]
I doubt this will get unmerged. They don't give a shit, it's clear.

**Comment by Al-1ce** [comment]
Hopefully this gets unmerged.

**Comment by Al-1ce** [comment]
> Code an "off" button so I can opt out of this mandatory 'merged while people were crashing to fire' shenanigans.


You don't seem to understand how this works. Either the entire server has fancy lighting or nobody does. It's all or nothing. I'm hoping it gets unmerged soon but I'm pretty doubtful.

**Pull request merged**